maggy/core/environment/base.py

```diff
@@ -0,0 +1,159 @@
+#
+#   Copyright 2021 Logical Clocks AB
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+import os
+
+import maggy.util as util
+
+class BaseEnv():
+    def __init__(self, *args):
+        self.log_dir = os.path.join(os.getcwd(), "experiment_log")
+        if not os.path.exists(self.log_dir):
+            os.mkdir(self.log_dir)
+        self.constants = []
+
+    def set_ml_id(self, app_id=0, run_id=0):
+        os.environ["ML_ID"] = str(app_id) + "_" + str(run_id)
+
+    def create_experiment_dir(self, app_id, run_id):
+        pass
+
+    def get_logdir(self, app_id, run_id):
+        return self.log_dir
+
+    def populate_experiment(
+        self,
+        model_name,
+        function,
+        type,
+        hp,
+        description,
+        app_id,
+        direction,
+        optimization_key,
+    ):
+        pass
+
+    def attach_experiment_xattr(self, exp_ml_id, experiment_json, command):
+        pass
+
+    def exists(self, hdfs_path, project=None):
+        return os.path.exists(hdfs_path)
+
+    def mkdir(self, hdfs_path, project=None):
+        pass
+
+    def isdir(self, dir_path):
+        return os.path.isdir(dir_path)
+
+    def ls(self, dir_path):
+        _, dirnames, filenames = next(os.walk(dir_path))
+
+        return dirnames + filenames
+
+    def delete(self, path, recursive=False):
+        if self.exists(path):
+            if os.path.isdir(path):
+                os.rmdir(path)
+            elif os.path.isfile(path):
+                os.remove(path)
+
+    def dump(self, data, hdfs_path):
+        head_tail = os.path.split(hdfs_path)
+        if not os.path.exists(head_tail[0]):
+            os.mkdir(head_tail[0])
+        file = self.open_file(hdfs_path, flags="w+")
+        file.write(data)
+
+    def get_ip_address(self):
+        sc = util.find_spark().sparkContext
+        return sc._conf.get("spark.driver.host")
+
+    def get_constants(self):
+        pass
+
+    def open_file(self, hdfs_path, project=None, flags="rw", buff_size=0):
```

Comment:
why `flags="rw"`?

Reply:
It's safer to just open in read-only on the default flags.